subprojects/base-services/src/main/java/org/gradle/internal/FileUtils.java

```diff
@@ -72,29 +93,17 @@ public static File assertInWindowsPathLengthLimitation(File file) {
      * @return the encompassing roots
      */
     public static Collection<? extends File> calculateRoots(Iterable<? extends File> files) {
-        List<File> roots = new LinkedList<File>();
-
-        files:
-        for (File file : files) {
-            File absoluteFile = file.getAbsoluteFile();
-            String path = file.getPath();
-            Iterator<File> rootsIterator = roots.iterator();
-
-            while (rootsIterator.hasNext()) {
-                File root = rootsIterator.next();
-                String rootPath = root.getPath();
-                if (doesPathStartWith(path, rootPath)) { // is lower than root
-                    continue files;
-                }
-
-                if (doesPathStartWith(rootPath, path)) { // is higher than root
-                    rootsIterator.remove();
-                }
+        List<File> roots = Lists.newArrayList(files);
+        Collections.sort(roots, FILE_SEGMENT_COMPARATOR);
+        File root = null;
+        for (Iterator<File> iterator = roots.iterator(); iterator.hasNext();) {
+            File next = iterator.next();
+            if (root != null && doesPathStartWith(next.getPath(), root.getPath())) {
+                iterator.remove();
```

Comment:
Adding to the end of an ArrayList with sufficient capacity is going to be ever slightly more efficient than removing from an ArrayList. The latter does not allocate more memory, but it involves an array copy. If it does not complicate the code too much, I'd suggest this change.